### PR TITLE
Feat/leveraged steth

### PR DESCRIPTION
**Pull request by Andreadinenno**
Implement a leveraged strategy on wstETH using the aave protocol. The vault essentially manages a CDP with `wstETH` collateral and `ETH` debt.

### Configuration: 
- `targetLTV`: the Loan to Value the vault looks to maintain 
- `maxLTV`: the maximum Loan to Value the vault can reach before triggering a rebalance 

### High level view of the flows: 

- `Deposit` -> takes wstETH from user and deposits into lending protocol
-  `Withdraw` -> if the amount requested to withdraw gest the vault above `maxLTV`, repay part of the debt (full repay if `totalAssets` withdraw) and withdraw that amount. If not, simply withdraw amount.
-  `Increase leverage` -> if vault LTV is lower than target, the vault does a loan of ETH, wraps into wstETH, deposits into CDP.
- `Reduce leverage` -> if vault LTV is higher than target, the vault does a flash loan of ETH, repay part of debt, withdraws wstETH, swaps for ETH (needed to repay the flashloan)

### CODE REVIEW

File: src/vault/adapter/lido/LeveragedWstETHAdapter.sol [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+        uint256 targetLTV_,
+        uint256 maxLTV_,
+        uint256 slippage_
+    ) external onlyOwner {
+        targetLTV = targetLTV_;
+        maxLTV = maxLTV_;
+
+        adjustLeverage();
+
+        slippage = slippage_;
```
Comment (0xruhum) [inline src/vault/adapter/lido/LeveragedWstETHAdapter.sol:453]: why do we adjust the leverage before updating the `slippage` parameter? Should add a small comment imo

File: src/vault/adapter/lido/LeveragedWstETHAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+        bool isFullWithdraw,
+        uint256 toWithdraw,
+        uint256 flashLoanDebt
+    ) internal {
+        address asset = asset();
+
+        // get flash loan amount of wstETH
+        uint256 flashLoanWstETHAmount = IwstETH(asset).getWstETHByStETH(
+            flashLoanDebt
+        );
```
Comment (0xruhum) [inline src/vault/adapter/lido/LeveragedWstETHAdapter.sol:300]: expects 1 WETH = 1 stETH. Not always the case.

File: src/vault/adapter/lido/LeveragedWstETHAdapter.sol [hunk 3]
```diff
@@ -27,7 +27,9 @@ contract LeveragedWstETHAdapter is AdapterBase, IFlashLoanReceiver {
     ILendingPool public lendingPool;
     IPoolAddressesProvider public poolAddressesProvider;
 
-    IWETH public weth;
+    IWETH public constant weth = IWETH(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+    address public constant stETH = address(0xae7ab96520DE3A18E5e111B5EaAb095312D7fE84);
+
     IERC20 public debtToken; // aave eth debt token
     IERC20 public interestToken; // aave awstETH
```
Comment (0xruhum) [inline src/vault/adapter/lido/LeveragedWstETHAdapter.sol:34]: the debt and interest tokens are also known, no?

Comment (Andreadinenno) [reply src/vault/adapter/lido/LeveragedWstETHAdapter.sol:34]: Yes, though the idea is to support both Aave and Spark by just changing the poolAddressesProvider address

Comment (0xruhum) [reply src/vault/adapter/lido/LeveragedWstETHAdapter.sol:34]: ok :+1:

**Pull request merged**